### PR TITLE
feat(treesitter): support self injection

### PR DESCRIPTION
**Pull request by ObserverOfTime**
See https://github.com/nvim-treesitter/nvim-treesitter/pull/3573#pullrequestreview-1132968510 for the motivation.

This seems to work and I'll try to add some actual tests later to make sure it does.

#### Example usage:

`c/injections.scm`
```scheme
;; preproc_arg gets injected as c
(preproc_arg) @self
```

`cpp/injections.scm`
```scheme
; inherits: c
;; preproc_arg gets injected as cpp
```

### CODE REVIEW

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 1]
```diff
@@ -388,7 +388,11 @@ function LanguageTree:_get_injections()
           -- Allows for other tags to be used in matches
         elseif string.sub(name, 1, 1) ~= '_' then
           if not lang then
-            lang = name
+            if name == 'self' then
+              lang = self:lang()
+            else
+              lang = name
+            end
```
Comment (lewis6991) [inline runtime/lua/vim/treesitter/languagetree.lua:395]:
```suggestion
            lang = name == 'self' and self:lang() or name
```

**Comment by ObserverOfTime** [comment]
The other special captures like `@combined` aren't uppercase either.

**Comment by clason** [comment]
Are capture names case sensitive always? (Highlight groups are not.)

**Comment by lewis6991** [comment]
Haven't checked, changing case was just one suggestion. The main point was doing something to better distinguish languages from meta stuff (bikeshedding aside).

We could even do something like `@meta.self` or `@_self`.

**Comment by clason** [comment]
> The main point was doing something to better distinguish languages from meta stuff (bikeshedding aside).

That is a good point (and also applies to `@conceal` and `@spell`).

I think I would prefer `@_self`, as "underscore means internal" is a more widely used pattern.

**Comment by ObserverOfTime** [comment]
`@_self` may be used in queries to match the `self` keyword in languages like Lua or Python.

**Comment by clason** [comment]
Ah, that's right, nvim-treesitter allows bypassing the Capture Police via `@_`.

**Comment by ObserverOfTime** [comment]
### Highlights

- `@conceal`
- `@spell`

### Injections

- `@language`
- `@content`
- `@combined`
- `@self` (?)

---

I think that's all of them.

**Comment by clason** [comment]
I'd hate to touch the special highlight captures; that would be (too) breaking.

I agree that making them stand out as special would be a good idea, but it's not enough benefit for triggering downstream breaking changes right now. Let's go with `@self` for now -- if somebody is foolish enough to create a SELF language, we'll just have to name the parser `selflang`.

**Comment by ObserverOfTime** [comment]
Not sure why the table order is messed up with `@self` while they're in fact the same regions as `@c`. :thinking: 
```
RUN      T3987 treesitter parser API when creating a language tree when parsing regions independently should inject a language: 5.73 ms OK
RUN      T3988 treesitter parser API when creating a language tree when parsing regions independently should inject the parent language with @self: 8.68 ms FAIL
test/functional/treesitter/parser_spec.lua:646: Expected objects to be the same.
Passed in:
(table: 0x7fb1d0ecb6b0) {
  [1] = {
    [1] = 0
    [2] = 0
    [3] = 7
    [4] = 0 }
 *[2] = {
   *[1] = 1
    [2] = 26
    [3] = 1
    [4] = 65 }
  [3] = {
    [1] = 2
    [2] = 29
    [3] = 2
    [4] = 68 }
  [4] = {
    [1] = 3
    [2] = 14
    [3] = 3
    [4] = 17 }
  [5] = {
    [1] = 4
    [2] = 15
    [3] = 4
    [4] = 18 }
  [6] = {
    [1] = 5
    [2] = 15
    [3] = 5
    [4] = 18 } }
Expected:
(table: 0x7fb1d092f3a8) {
  [1] = {
    [1] = 0
    [2] = 0
    [3] = 7
    [4] = 0 }
 *[2] = {
   *[1] = 3
    [2] = 14
    [3] = 3
    [4] = 17 }
  [3] = {
    [1] = 4
    [2] = 15
    [3] = 4
    [4] = 18 }
  [4] = {
    [1] = 5
    [2] = 15
    [3] = 5
    [4] = 18 }
  [5] = {
    [1] = 1
    [2] = 26
    [3] = 1
    [4] = 65 }
  [6] = {
    [1] = 2
    [2] = 29
    [3] = 2
    [4] = 68 } }
```

**Comment by binyomen** [comment]
There is already a [self language](https://en.m.wikipedia.org/wiki/Self_(programming_language)), and has been since the 80s ☹️. It's one of the early object oriented languages. I couldn't find a tree-sitter parser for it and I don't know how commonly used it still is (latest release in 2017), but that's something to keep in mind.

I think naming a future parser selflang makes sense.

**Comment by theHamsta** [comment]
We could implement this feature in nvim-treesitter with a custom predicate. This would maybe be less invasive than this change in core.

**Comment by clason** [comment]
As highlighting is now "owned" by core, this should be done here unless there's a plan for iterating a lot.

**Comment by ObserverOfTime** [comment]
We can always name the capture something else, like `@parent`.
Surely there's no "parent" language…right?

**Comment by BaraTudor2025** [comment]
What about `@nvim.self`, `@nvim.conceal`, etc. ?  It's pretty obvious that these captures are internal or specific to neovim. 
Also, `@nvim` could be used as a namespace for future features without worrying about conflicts with other queries.

**Comment by lewis6991** [comment]
> What about `@nvim.self`, `@nvim.conceal`, etc. ? It's pretty obvious that these captures are internal or specific to neovim. Also, `@nvim` could be used as a namespace for future features without worrying about conflicts with other queries.

I like this. I think we should do it.

**Comment by ObserverOfTime** [comment]
It's a fine idea but it would require changing the existing captures both here and in every treesitter-related plugin.

**Comment by ObserverOfTime** [comment]
So, should I use `@nvim.self` for the capture?
Also, can anyone help me figure out [this](https://github.com/neovim/neovim/pull/20712#issuecomment-1282445137) issue?

**Comment by clason** [comment]
I think we want to bank this until we can coordinate this with Helix (as part of the grand Conjunction of the Queries).

**Comment by lewis6991** [comment]
Due to https://github.com/neovim/neovim/pull/22518. We may need to rethink this.

**Comment by clason** [comment]
Injection format PR is merged; going forward this is the only supported format (so new features will not be supported by the old format).

This means we could use `(#set! injection.language.self)` and avoid conflict with any language name.

Although I'm skeptical that this actually has a pressing use case; the two injection files mentioned in the linked PR are tiny (two queries each), so combining them doesn't really offset the cost of the new feature?

**Comment by ObserverOfTime** [comment]
> Although I'm skeptical that this actually has a pressing use case; the two injection files mentioned in the linked PR are tiny (two queries each), so combining them doesn't really offset the cost of the new feature?

It's necessary for nvim-treesitter/nvim-treesitter#3573.

**Comment by clason** [comment]
Could you summarize the key issue in this PR (ideally commit) description? Otherwise it becomes hard to discuss it here.

If I understand it correctly, the _actual_ issue (which is not the example usage you gave!) is that you have the injection chain
```
lang A -> lang B -> lang A
```
but the queries for lang B don't know where it was injected into?

Then `injection.language.self` is wrong (or at least misleading); should be `injection.language.parent` instead.

**Comment by ObserverOfTime** [comment]
Let's say we have languages `A` and `B` where `B` extends `A` and
that there's a `foo` node (e.g. preprocessor argument or string
interpolation) which requires an injection of the language itself.
The issue here is that `B` cannot simply inherit from `A` because
`foo` would be injected as `A` when it should be injected as `B`.

Any chance the new injection method has already solved this?

**Comment by clason** [comment]
> The issue here is that B cannot simply inherit from A because
foo would be injected as A when it should be injected as B.

Then you don't inherit the injections, period; for the given example (c and cpp) where it's just a single query, I don't see the need to add new features to save two or three lines of runtime file?


> Any chance the new injection method has already solved this?

I don't see how? That's just about how you specify the injections.

**Comment by ObserverOfTime** [comment]
> Then you don't inherit the injections, period; for the given example (c and cpp) where it's just a single query, I don't see the need to add new features to save two or three lines of runtime file?

nvim-treesitter/nvim-treesitter#3573 adds a lot more queries to them. :)

> I don't see how? That's just about how you specify the injections.

Perhaps by allowing inherited queries to be overwritten.

**Comment by ObserverOfTime** [comment]
I've updated this for the new format.
```clojure
(preproc_def (preproc_arg) @injection.content (#set! injection.self))
```

**Comment by lewis6991** [comment]
I'm not too keen on adding this unless we can get this recognized by upstream.

**Comment by amaanq** [comment]
I wouldn't mind adding that upstream

**Comment by amaanq** [comment]
https://github.com/tree-sitter/tree-sitter/pull/2529

**Comment by ObserverOfTime** [comment]
Superseded by #24831.
